README
Simplize using Fusionchartfree for ruby on rails


First you should generate swf files

rails g easyfsf


and then you can render chart by using helper method

example


<%= render_sm_chart("a", "Pie2D", ['a','b','c'], [1,2,3]) %>
<%= render_sm_chart("b", "Column3D", ['a','b','c'], [{:seriesName => "se1", :color => "FF0000",:value_list => [1,2,3]}, {:seriesName => "se2", :color => "F660AB", :value_list => [3,4,5]}]) %>


if you want better example including some images, visit  http://zerohun.wordpress.com/2012/01/06/easyfsf/

rubydoc
http://rubydoc.info/github/zerohun/Easyfsf/master/frames



CHANGELOG.md
## v0.0.1

*initial release


spec/spec_helper.rb
require 'easyfsf'
include Easyfsf
require "active_support/core_ext/object/blank"
require "active_support/core_ext/string/output_safety"

spec/easyfsf/easyfsf_spec.rb
# encoding: utf-8
require 'spec_helper'
describe Easyfsf do
  describe "render charts" do
    it "render simple single chart" do
      render_sm_chart("a", "Pie2D", ['a','b','c'], [1,2,3]).should eq("<div id='aDiv' class='fusionchart'>\n\t\t\t\tChart.\n\t\t<fusionchartfree swf='/FusionChartsFree/Charts/FCF_Pie2D.swf' chart_id='a' chart_width='700' chart_height='500' chart_xml='<graph caption=\"\" subcaption=\"\" lineThickness=\"2\" showValues=\"0\" formatNumberScale=\"0\" anchorRadius=\"1\" divLineAlpha=\"20\" divLineColor=\"CC3300\" showAlternateHGridColor=\"1\" alternateHGridColor=\"CC3300\" rotateNames=\"1\" shadowAlpha=\"40\" numvdivlines=\"5\" chartLeftMargin=\"5\" chartRightMargin=\"5\" chartTopMargin=\"5\" chartBottomMargin=\"5\" bgColor=\"FFFFFF\" alternateHGridAlpha=\"5\" limitsDecimalPrecision=\"0\" divLineDecimalPrecision=\"0\" decimalPrecision=\"0\" yAxisMaxValue=\"10\"><set name=\"a\" value=\"1\" color=\"\"/><set name=\"b\" value=\"2\" color=\"\"/><set name=\"c\" value=\"3\" color=\"\"/></graph>' /></div>")
    end
    
    it "render simple multie chart" do
      render_sm_chart("b", "Column3D", ['a','b','c'], [{:seriesName => "se1", :color => "FF0000",:value_list => [1,2,3]}, {:seriesName => "se2", :color => "F660AB", :value_list => [3,4,5]}]).should eq("<div id='bDiv' class='fusionchart'>\n\t\t\t\tChart.\n\t\t<fusionchartfree swf='/FusionChartsFree/Charts/FCF_MSColumn3D.swf' chart_id='b' chart_width='700' chart_height='500' chart_xml='<graph caption=\"\" subcaption=\"\" lineThickness=\"2\" showValues=\"0\" formatNumberScale=\"0\" anchorRadius=\"1\" divLineAlpha=\"20\" divLineColor=\"CC3300\" showAlternateHGridColor=\"1\" alternateHGridColor=\"CC3300\" rotateNames=\"1\" shadowAlpha=\"40\" numvdivlines=\"5\" chartLeftMargin=\"5\" chartRightMargin=\"5\" chartTopMargin=\"5\" chartBottomMargin=\"5\" bgColor=\"FFFFFF\" alternateHGridAlpha=\"5\" limitsDecimalPrecision=\"0\" divLineDecimalPrecision=\"0\" decimalPrecision=\"0\" yAxisMaxValue=\"10\"><categories><category name=\"a\"/><category name=\"b\"/><category name=\"c\"/></categories><dataset seriesName=\"se1\" color=\"FF0000\"><set value=\"1\"/><set value=\"2\"/><set value=\"3\"/></dataset><dataset seriesName=\"se2\" color=\"F660AB\"><set value=\"3\"/><set value=\"4\"/><set value=\"5\"/></dataset></graph>' /></div>")
    end
  end
end

lib/generators/easyfsf/templates/app/assets/javascripts/fusion_charts.js
/**
 * FusionCharts: Flash Player detection and Chart embedding.
 * Version 1.2.3F ( 22 November 2008) - Specialized for FusionChartsFREE 
 *                     Checking Flash Version >=6 and added updateChartXML() for FREE Charts.
 * Version: 1.2.3 (1st September, 2008) - Added Fix for % and & characters, scaled dimensions, fixes in to properly handling of double quotes and single quotes in setDataXML() function.
 * Version: 1.2.2 (10th July, 2008) - Added Fix for % scaled dimensions, fixes in setDataXML() and setDataURL() functions
 * Version: 1.2.1 (21st December, 2007) - Added setting up Transparent/opaque mode: setTransparent() function 
 * Version: 1.2 (1st November, 2007) - Added FORM fixes for IE 
 * Version: 1.1 (29th June, 2007) - Added Player detection, New conditional fixes for IE
 *
 * Morphed from SWFObject (http://blog.deconcept.com/swfobject/) under MIT License:
 * http://www.opensource.org/licenses/mit-license.php
 *
 */
if(typeof infosoftglobal == "undefined") var infosoftglobal = new Object();
if(typeof infosoftglobal.FusionChartsUtil == "undefined") infosoftglobal.FusionChartsUtil = new Object();
infosoftglobal.FusionCharts = function(swf, id, w, h, debugMode, registerWithJS, c, scaleMode, lang, detectFlashVersion, autoInstallRedirect){
    if (!document.getElementById) { return; }
    
    //Flag to see whether data has been set initially
    this.initialDataSet = false;
    
    //Create container objects
this.params = new Object();
this.variables = new Object();
this.attributes = new Array();

//Set attributes for the SWF
if(swf) { this.setAttribute('swf', swf); }
if(id) { this.setAttribute('id', id); }

w=w.toString().replace(/\%$/,"%25");
if(w) { this.setAttribute('width', w); }
h=h.toString().replace(/\%$/,"%25");
if(h) { this.setAttribute('height', h); }


//Set background color
if(c) { this.addParam('bgcolor', c); }

//Set Quality
this.addParam('quality', 'high');

//Add scripting access parameter
this.addParam('allowScriptAccess', 'always');

//Pass width and height to be appended as chartWidth and chartHeight
this.addVariable('chartWidth', w);
this.addVariable('chartHeight', h);

//Whether in debug mode
debugMode = debugMode ? debugMode : 0;
this.addVariable('debugMode', debugMode);
//Pass DOM ID to Chart
this.addVariable('DOMId', id);
//Whether to registed with JavaScript
registerWithJS = registerWithJS ? registerWithJS : 0;
this.addVariable('registerWithJS', registerWithJS);

//Scale Mode of chart
scaleMode = scaleMode ? scaleMode : 'noScale';
this.addVariable('scaleMode', scaleMode);

//Application Message Language
lang = lang ? lang : 'EN';
this.addVariable('lang', lang);

//Whether to auto detect and re-direct to Flash Player installation
this.detectFlashVersion = detectFlashVersion?detectFlashVersion:1;
this.autoInstallRedirect = autoInstallRedirect?autoInstallRedirect:1;

//Ger Flash Player version 
this.installedVer = infosoftglobal.FusionChartsUtil.getPlayerVersion();

if (!window.opera && document.all && this.installedVer.major > 7) {
// Only add the onunload cleanup if the Flash Player version supports External Interface and we are in IE
infosoftglobal.FusionCharts.doPrepUnload = true;
}
}

infosoftglobal.FusionCharts.prototype = {
setAttribute: function(name, value){
this.attributes[name] = value;
},
getAttribute: function(name){
return this.attributes[name];
},
addParam: function(name, value){
this.params[name] = value;
},
getParams: function(){
return this.params;
},
addVariable: function(name, value){
this.variables[name] = value;
},
getVariable: function(name){
return this.variables[name];
},
getVariables: function(){
return this.variables;
},
getVariablePairs: function(){
var variablePairs = new Array();
var key;
var variables = this.getVariables();
for(key in variables){
variablePairs.push(key +"="+ variables[key]);
}
return variablePairs;
},
getSWFHTML: function() {
var swfNode = "";
if (navigator.plugins && navigator.mimeTypes && navigator.mimeTypes.length) { 
// netscape plugin architecture
swfNode = '<embed wmode="transparent" type="application/x-shockwave-flash" src="'+ this.getAttribute('swf') +'" width="'+ this.getAttribute('width') +'" height="'+ this.getAttribute('height') +'"  ';
swfNode += ' id="'+ this.getAttribute('id') +'" name="'+ this.getAttribute('id') +'" ';
var params = this.getParams();
 for(var key in params){ swfNode += [key] +'="'+ params[key] +'" '; }
var pairs = this.getVariablePairs().join("&");
 if (pairs.length > 0){ swfNode += 'flashvars="'+ pairs +'"'; }
swfNode += '/>';
} else { // PC IE
swfNode = '<object id="'+ this.getAttribute('id') +'" classid="clsid:D27CDB6E-AE6D-11cf-96B8-444553540000" width="'+ this.getAttribute('width') +'" height="'+ this.getAttribute('height') +'">';
swfNode += '<param name="movie" value="'+ this.getAttribute('swf') +'" />';
swfNode += '<param name="wmode" value="transparent">' 
var params = this.getParams();
for(var key in params) {
 swfNode += '<param name="'+ key +'" value="'+ params[key] +'" />';
}
var pairs = this.getVariablePairs().join("&");
if(pairs.length > 0) {swfNode += '<param name="flashvars" value="'+ pairs +'" />';}
swfNode += "</object>";
}
return swfNode;
},
setDataURL: function(strDataURL){
//This method sets the data URL for the chart.
//If being set initially
if (this.initialDataSet==false){
this.addVariable('dataURL',strDataURL);
//Update flag
this.initialDataSet = true;
}else{
//Else, we update the chart data using External Interface
//Get reference to chart object
var chartObj = infosoftglobal.FusionChartsUtil.getChartObject(this.getAttribute('id'));

if (!chartObj.setDataURL)
{
__flash__addCallback(chartObj, "setDataURL");
}

chartObj.setDataURL(strDataURL);
}
},
//This function :
//fixes the double quoted attributes to single quotes
//Encodes all quotes inside attribute values
//Encodes % to %25 and & to %26;
encodeDataXML: function(strDataXML){

var regExpReservedCharacters=["\\$","\\+"];
var arrDQAtt=strDataXML.match(/=\s*\".*?\"/g);
if (arrDQAtt){
for(var i=0;i<arrDQAtt.length;i++){
var repStr=arrDQAtt[i].replace(/^=\s*\"|\"$/g,"");
repStr=repStr.replace(/\'/g,"%26apos;");
var strTo=strDataXML.indexOf(arrDQAtt[i]);
var repStrr="='"+repStr+"'";
var strStart=strDataXML.substring(0,strTo);
var strEnd=strDataXML.substring(strTo+arrDQAtt[i].length);
var strDataXML=strStart+repStrr+strEnd;
}
}

strDataXML=strDataXML.replace(/\"/g,"%26quot;");
strDataXML=strDataXML.replace(/%(?![\da-f]{2}|[\da-f]{4})/ig,"%25");
strDataXML=strDataXML.replace(/\&/g,"%26");

return strDataXML;

},
setDataXML: function(strDataXML){
//If being set initially
if (this.initialDataSet==false){
//This method sets the data XML for the chart INITIALLY.
this.addVariable('dataXML',this.encodeDataXML(strDataXML));
//Update flag
this.initialDataSet = true;
}else{
//Else, we update the chart data using External Interface
//Get reference to chart object
var chartObj = infosoftglobal.FusionChartsUtil.getChartObject(this.getAttribute('id'));
chartObj.setDataXML(strDataXML);
}
},
setTransparent: function(isTransparent){
//Sets chart to transparent mode when isTransparent is true (default)
//When no parameter is passed, we assume transparent to be true.
if(typeof isTransparent=="undefined") {
isTransparent=true;
}
//Set the property
if(isTransparent)
this.addParam('WMode', 'transparent');
else
this.addParam('WMode', 'Opaque');
},
remove: function(chartId){},

render: function(elementId){
//First check for installed version of Flash Player - we need a minimum of 6
if((this.detectFlashVersion==1) && (this.installedVer.major < 6)){
if (this.autoInstallRedirect==1){
//If we can auto redirect to install the player?
var installationConfirm = window.confirm("You need Adobe Flash Player 6 (or above) to view the charts. It is a free and lightweight installation from Adobe.com. Please click on Ok to install the same.");
if (installationConfirm){
window.location = "http://www.adobe.com/shockwave/download/download.cgi?P1_Prod_Version=ShockwaveFlash";
}else{
return false;
}
}else{
//Else, do not take an action. It means the developer has specified a message in the DIV (and probably a link).
//So, expect the developers to provide a course of way to their end users.
//window.alert("You need Adobe Flash Player 8 (or above) to view the charts. It is a free and lightweight installation from Adobe.com. ");
return false;
}
}else{
//Render the chart
var n = (typeof elementId == 'string') ? document.getElementById(elementId) : elementId;
n.innerHTML = this.getSWFHTML();

//Added <FORM> compatibility
//Check if it's added in Mozilla embed array or if already exits 
if(!document.embeds[this.getAttribute('id')] && !window[this.getAttribute('id')])
      window[this.getAttribute('id')]=document.getElementById(this.getAttribute('id')); 
//or else document.forms[formName/formIndex][chartId]
return true;
}
}
}

/* ---- detection functions ---- */
infosoftglobal.FusionChartsUtil.getPlayerVersion = function(){
var PlayerVersion = new infosoftglobal.PlayerVersion([0,0,0]);
if(navigator.plugins && navigator.mimeTypes.length){
var x = navigator.plugins["Shockwave Flash"];
if(x && x.description) {
PlayerVersion = new infosoftglobal.PlayerVersion(x.description.replace(/([a-zA-Z]|\s)+/, "").replace(/(\s+r|\s+b[0-9]+)/, ".").split("."));
}
}else if (navigator.userAgent && navigator.userAgent.indexOf("Windows CE") >= 0){ 
//If Windows CE
var axo = 1;
var counter = 3;
while(axo) {
try {
counter++;
axo = new ActiveXObject("ShockwaveFlash.ShockwaveFlash."+ counter);
PlayerVersion = new infosoftglobal.PlayerVersion([counter,0,0]);
} catch (e) {
axo = null;
}
}
} else { 
// Win IE (non mobile)
// Do minor version lookup in IE, but avoid Flash Player 6 crashing issues
try{
var axo = new ActiveXObject("ShockwaveFlash.ShockwaveFlash.7");
}catch(e){
try {
var axo = new ActiveXObject("ShockwaveFlash.ShockwaveFlash.6");
PlayerVersion = new infosoftglobal.PlayerVersion([6,0,21]);
axo.AllowScriptAccess = "always"; // error if player version < 6.0.47 (thanks to Michael Williams @ Adobe for this code)
} catch(e) {
if (PlayerVersion.major == 6) {
return PlayerVersion;
}
}
try {
axo = new ActiveXObject("ShockwaveFlash.ShockwaveFlash");
} catch(e) {}
}
if (axo != null) {
PlayerVersion = new infosoftglobal.PlayerVersion(axo.GetVariable("$version").split(" ")[1].split(","));
}
}
return PlayerVersion;
}
infosoftglobal.PlayerVersion = function(arrVersion){
this.major = arrVersion[0] != null ? parseInt(arrVersion[0]) : 0;
this.minor = arrVersion[1] != null ? parseInt(arrVersion[1]) : 0;
this.rev = arrVersion[2] != null ? parseInt(arrVersion[2]) : 0;
}
// ------------ Fix for Out of Memory Bug in IE in FP9 ---------------//
/* Fix for video streaming bug */
infosoftglobal.FusionChartsUtil.cleanupSWFs = function() {
var objects = document.getElementsByTagName("OBJECT");
for (var i = objects.length - 1; i >= 0; i--) {
objects[i].style.display = 'none';
for (var x in objects[i]) {
if (typeof objects[i][x] == 'function') {
objects[i][x] = function(){};
}
}
}
}
// Fixes bug in fp9
if (infosoftglobal.FusionCharts.doPrepUnload) {
if (!infosoftglobal.unloadSet) {
infosoftglobal.FusionChartsUtil.prepUnload = function() {
__flash_unloadHandler = function(){};
__flash_savedUnloadHandler = function(){};
window.attachEvent("onunload", infosoftglobal.FusionChartsUtil.cleanupSWFs);
}
window.attachEvent("onbeforeunload", infosoftglobal.FusionChartsUtil.prepUnload);
infosoftglobal.unloadSet = true;
}
}
/* Add document.getElementById if needed (mobile IE < 5) */
if (!document.getElementById && document.all) { document.getElementById = function(id) { return document.all[id]; }}
/* Add Array.push if needed (ie5) */
if (Array.prototype.push == null) { Array.prototype.push = function(item) { this[this.length] = item; return this.length; }}

/* Function to return Flash Object from ID */
infosoftglobal.FusionChartsUtil.getChartObject = function(id)
{
  var chartRef=null;
  if (navigator.appName.indexOf("Microsoft Internet")==-1) {
    if (document.embeds && document.embeds[id])
      chartRef = document.embeds[id]; 
else
chartRef  = window.document[id];
  }
  else {
    chartRef = window[id];
  }
  if (!chartRef)
chartRef  = document.getElementById(id);
  
  return chartRef;
}
/*
 Function to update chart's data at client side (FOR FusionCharts vFREE and 2.x
*/
infosoftglobal.FusionChartsUtil.updateChartXML = function(chartId, strXML){
//Get reference to chart object
var chartObj = infosoftglobal.FusionChartsUtil.getChartObject(chartId);
//Set dataURL to null
chartObj.SetVariable("_root.dataURL","");
//Set the flag
chartObj.SetVariable("_root.isNewData","1");
//Set the actual data
chartObj.SetVariable("_root.newData",strXML);
//Go to the required frame
chartObj.TGotoLabel("/", "JavaScriptHandler"); 
}

/* Aliases for easy usage */
var getChartFromId = infosoftglobal.FusionChartsUtil.getChartObject;
var updateChartXML = infosoftglobal.FusionChartsUtil.updateChartXML;
var FusionCharts = infosoftglobal.FusionCharts;

/* 
 * <fusionchartfree> 엘리먼트를 파싱해서 attibute에 맞춰 차트를  생성합니다.
 */

function render_element(element){
        var $chart = $(element)
        var chart_id = $chart.attr("chart_id");
        var chart_swf = $chart.attr("swf")
        var chart_width = $chart.attr("chart_width");
        var chart_height = $chart.attr("chart_height");
        var chart_xml = $chart.attr("chart_xml");
        var chart_object = new FusionCharts(chart_swf, chart_id, chart_width, chart_height, "0", "0");
        var div_id_ar = []
        div_id_ar[div_id_ar.length] = chart_id
        div_id_ar[div_id_ar.length] = 'Div'
        var div_id_str = div_id_ar.join("")
        chart_object.setDataXML(chart_xml);
        chart_object.render(div_id_str)
}
$(document).ready(function(){
    var chart_list = $("fusionchartfree");
    var chart_options = {};

    chart_list.each(function(index, element){
        var chart_id = $(element).attr('id');
        var chart_xml =  $(element).attr('chart_xml');
        var attributes = chart_xml.attributes
        chart_options[chart_id] = ""
        for(var key in attributes){
            chart_options[chart_id] +=  ( attributes[key].name + "="+ attributes[key].value + ';')
        }
        render_element(element);
    });
    jQuery.data(document.body, 'fusion chart free option', chart_options);
});





lib/easyfsf.rb
require "easyfsf/version"
require "easyfsf/application_helper" if defined? Rails
require "builder"


module Easyfsf

  #To render single chart or multi chart, call <tt>render_sm_chart</tt>
  #Depends on parameter it render single data chart or multi data chart
  #You can use this method as helper
  #
  #<%= render_sm_chart("a", "Pie2D", ['a','b','c'], [1,2,3]) %>
  #
  #<%= render_sm_chart("b", "Column3D", ['a','b','c'], [{:seriesName => "se1", :color => "FF0000",:value_list => [1,2,3]}, {:seriesName => "se2", :color => "F660AB", :value_list => [3,4,5]}]) %>
  #<%= render_sm_chart("b", "Column3D", ['a','b','c'], [{:seriesName => "se1", :color => "FF0000",:value_list => [1,2,3]}, {:seriesName => "se2", :color => "F660AB", :value_list => [3,4,5]}],{}, 500, 400, true) %>
  #
  #for detail infomation visit http://zerohun.wordpress.com/2012/01/06/easyfsf/
  #for the options parameter visit http://docs.fusioncharts.com/charts/
  
  def render_sm_chart(chart_id, chart_type, category, datasets, options = {}, chart_width = 700, chart_height = 500, show_total = false)
    
    options = options.clone if options.present?
    category = FCFInitializer.standardize_category(category)
    datasets = FCFInitializer.standardize_datasets(datasets)
    
    if options[:skipCategory].present?
      category = FCFInitializer.skip_category(category, 10)
      options.delete :skipCategory
    end
    
    isMultiChart = datasets.length > 1
    chart_swf = FCFInitializer.get_swf_file_name(chart_type, isMultiChart)
    options = FCFInitializer.get_default_option.merge(options)
    chart_html = ""
    if show_total == true
      chart_html << "<div id=#{chart_id}Div_total><br />"
      datasets.each do |dataset|
        total = 0
        dataset[:value_list].each do |value|
          total += value
        end
        chart_html << "<p>Total #{dataset[:seriesName]} : #{total}</p>"
      end
      chart_html << "<br /></div>"
    end
    xmlString = FCFChartXmlBuilder.smchart_xml(category, datasets, options)
    chart_html << FCFRenderer.render_chart("/FusionChartsFree/Charts/#{chart_swf}", xmlString, chart_id, chart_width, chart_height)
    return chart_html.html_safe
  end
  

  
  module FCFInitializer
    #default optinoss
    DEFAULT_OPTIONS = {
          :caption                   => "",
          :subcaption                => "",
          :lineThickness             => "5",
          :showValues                => "0",
          :formatNumberScale         => "0", 
          :anchorRadius              => "5", 
          :divLineAlpha              => "20" ,
          :divLineColor              => "CC3300", 
          :showAlternateHGridColor   => "1", 
          :alternateHGridColor       => "CC3300",
          :rotateNames               => "1",
          :shadowAlpha               => "40",  
          :numvdivlines              => "5", 
          :chartLeftMargin           => "5", 
          :chartRightMargin          => "5", 
          :chartTopMargin            => "5", 
          :chartBottomMargin         => "5", 
          :bgColor                   => "FFFFFF", 
          :alternateHGridAlpha       => "5", 
          :limitsDecimalPrecision    => '0', 
          :divLineDecimalPrecision   => '0',
          :decimalPrecision          => "0",
          :yAxisMaxValue             => "10",
          :lineThickness             => "2",
          :anchorRadius              => "1"
    }      
        


    #to file swf files
    #chart difinition for rendering single chart
    ONLY_SINGLE_CHART = ["Pie2D", "Pie3D", "Funnel"]
    
    #chart difinition for rendering multi chart
    ONLY_MULTI_CHART = ["StackedArea2D", "StackedBar2D", "StackedColumn2D", 
                        "StackedColumn3D", "MSColumn2DLineDY", "MSColumn3DLineDY"]
    
    #chart difinition for rendering single and multi chart
    BOTH_CAPABLE_CHART = ["Line", "Area2D", "Bar2D", "Column2D", "Column3D"]    
    @chartIdList = []
    
    def self.get_default_option
      DEFAULT_OPTIONS
    end
    

    def self.get_swf_file_name(chart_type, isMulti)
      if ONLY_SINGLE_CHART.include? chart_type 
        if isMulti
          raise ArgumentError, "#{chart_type} can draw only single chart data but it's multichart data"
        else
          return "FCF_#{chart_type}.swf"
        end
      end
      if ONLY_MULTI_CHART.include? chart_type
        if isMulti
          return "FCF_#{chart_type}.swf"
        else
          raise ArgumentError, "#{chart_type} can draw only multichart data but it's singlechart data"
        end
      end
      if BOTH_CAPABLE_CHART.include? chart_type
        if isMulti
          return "FCF_MS#{chart_type}.swf"
        else
          return "FCF_#{chart_type}.swf"
        end      
      end
      raise ArgumentError, "there is no #{chart_type} type"   
    end
    

    def self.standardize_category(category)
      case check_category_params(category)
      when :STRING_ARRAY  
        return {:value_list => category.map {|c| c.to_s}}
      when :HASH
        return category.clone
      else
        raise ArgumentError, 
              "First parameter should be Hash, Array of String or Array of something that can be converted to String"  
      end        
    end
      

    def self.standardize_datasets(datasets)
      case check_datasets_params(datasets)
      when :NUM_ARRAY
        return [{:value_list => datasets}]
      when :HASH_ARRAY
        clonedDataSets = []
        datasets.each do |dataset|
          clonedDataSets << dataset.clone
        end
        return clonedDataSets
      when :HASH
        return [datasets.clone]
      else
        raise ArgumentError, 
          "parameters should be Hash, Array of Hash , Array of Fixum or Array of Float"             
      end
    end
    def self.check_category_params(category)
      if category.is_a? Array
        return :STRING_ARRAY if category[0].respond_to?(:to_s)
      end
      if category.is_a? Hash
        return :HASH         if category[:value_list]
      end
      return :SOMETHING_ELSE
    end
            
    def self.check_datasets_params(datasets)
      if datasets.is_a? Array 
        return :NUM_ARRAY  if datasets[0].is_a? Fixnum or datasets[0].is_a? Float
        if datasets[0].is_a? Hash
          datasets.each_with_index do |dataset, index|
            raise ArgumentError, 
                "There is a nil value_list in datasets(name: #{dataset[:seriesName]} ,index:#{index})" if dataset[:value_list].blank?
          end
          return :HASH_ARRAY
        end
      end
      if datasets.is_a? Hash
        return :HASH if datasets[:value_list].present?
      end
      return :SOMETHING_ELSE
    end    
    
    #skipSize에 맞춰 category의 value_list를 빈 문자열로 교체합니다.
    def self.skip_category(category, skipSize)
      newValueList = category[:value_list].each_with_index.map{ |value,index|
        next value if (index+1)%(skipSize+1) == 1
        next ""
        }
      return category.merge(:value_list => newValueList)
    end
  end
  
  module FCFRenderer
    #차트의 xml 데이타를 <fusionchatfree />형식으로 나타내어서 javascript 코드에서 파싱해서
    #차트를 그려줄 수 있게끔 합니다. 
    def self.render_chart(chart_swf, chart_xml, chart_id , chart_width, chart_height) 
      chart_width=chart_width.to_s
      chart_height=chart_height.to_s
      htmlString = ""
      htmlString << "<div id='#{chart_id}Div' class='fusionchart'>\n\t\t\t\tChart.\n\t\t"
      htmlString << "<fusionchartfree swf='#{chart_swf}' chart_id='#{chart_id}' chart_width='#{chart_width}' chart_height='#{chart_height}' chart_xml='#{chart_xml}' />"
      htmlString << "</div>"
      return htmlString
    end    
  end    
  
  module FCFChartXmlBuilder
    
    # for randomize color
    @FC_ColorCounter=0;
    @arr_FCColors=["1941A5", "AFD8F8", "F6BD0F", "8BBA00", "A66EDD", "F984A1", 
                    "CCCC00", "999999", "0099CC", "FF0000", "006F00", "0099FF", 
                    "FF66CC", "669966", "7C7CB4", "FF9933", "9900FF", "99FFCC", 
                    "CCCCFF", "669900"]
                    
    # Generate xml for chart
    # examples
    #
    # for single charts
    #<graph  yAxisName='Sales Figure' 
    #                caption='Top 5 Sales Person' 
    #                numberPrefix='$' 
    #                decimalPrecision='1' 
    #                divlinedecimalPrecision='0' 
    #                limitsdecimalPrecision='0'>
    #        <set name='Alex' value='25000' color='AFD8F8'/> 
    #        <set name='Mark' value='35000' color='F6BD0F'/> 
    #        <set name='John' value='31300' color='008E8E'/> 
    #
    #</graph>
    #
    # for multi charts
    #<graph  yAxisName='Sales Figure' 
    #                caption='Top 5 Sales Person' 
    #                numberPrefix='$' 
    #                decimalPrecision='1' 
    #                divlinedecimalPrecision='0' 
    #                limitsdecimalPrecision='0'>
    #        <categories>
    #                <category name="Jan"/> 
    #                <category name="Feb"/> 
    #                <category name="Apr"/> 
    #        </categories>
    #        <dataset seriesName="Current Year" color="A66EDD">
    #                <set value="1127654"/> 
    #                <set value="1226234"/> 
    #                <set value="1311565"/> 
    #
    #        </dataset>
    #        <dataset seriesName="Previous Year" color="F6BD0F">
    #                <set value="927654"/> 
    #                <set value="1126234"/> 
    #                <set value="1111565"/> 
    #
    #        </dataset>
    #</graph>    
    
    def self.smchart_xml(category, datasets, options = {})
      if options[:sectionSize].present?
        category[:value_list] = erase_sections(category[:value_list], options[:sectionSize])
        options.delete :sectionSize
      end
      
      xml = ::Builder::XmlMarkup.new(:indent=>0) 
      #pOptions[:caption] = getCaption(valueDatas)
      xml.graph(options) do
        if datasets.length == 1
          category[:value_list].length.times do |index|
            dataset = datasets[0]
            color = category[:color_list].blank? ? "":category[:color_list][index]
            xml.set(:name  => category[:value_list][index], 
                    :value => dataset[:value_list][index],
                    :color => color)
          end
        else
          xml.categories do
            category[:value_list].length.times do |index|
              xml.category(:name  => category[:value_list][index])
            end          
          end
          datasets.each do |dataset|
            dataset[:color] = dataset[:color].blank? ? get_FC_color : dataset[:color] 
            valueList = dataset[:value_list].clone
            dataset.delete :value_list
            xml.dataset(dataset) do
              valueList.each do |value|
                xml.set(:value => value)
              end
            end
          end
        end 
      end
      xmlString = xml.to_s
      return xmlString[0..xmlString.length - 8]            
    end
    


    def self.get_FC_color 
      #Update index
      @FC_ColorCounter=@FC_ColorCounter+1
      counter = @FC_ColorCounter % (@arr_FCColors.size)
      #Return color
      return @arr_FCColors[counter]
    end     
  end

end


